Dev: allow tests to pass, even when not all tool dependencies are installed

### Is your feature request related to a problem? Please describe.

Running

```console
$ git clone https://github.com/goreleaser/goreleaser.git
...
$ cd goreleaser
$ go test ./...
```

on a newly-installed machine fails due to multiple missing binaries and services, e.g.:

```
2024/08/27 19:12:44 Get "http://unix.sock/_ping": dial unix /var/run/docker.sock: connect: no such file or directory
FAIL	github.com/goreleaser/goreleaser/v2/internal/pipe/blob	0.020s
...
  • nix-prefetch-url is not available
  • nix-prefetch-url is not available
--- FAIL: TestSkip (0.00s)
    --- FAIL: TestSkip/nix-all-good (0.00s)
        nix_test.go:38: 
            	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/nix/nix_test.go:38
            	Error:      	Should be false
            	Test:       	TestSkip/nix-all-good
  • fake-nix-prefetch-url is not available
  • nix-prefetch-url is not available
--- FAIL: TestPrefetcher (0.00s)
    --- FAIL: TestPrefetcher/prefetch (0.00s)
        --- FAIL: TestPrefetcher/prefetch/publish (0.00s)
            --- FAIL: TestPrefetcher/prefetch/publish/valid (0.00s)
                nix_test.go:66: 
                    	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/nix/nix_test.go:66
                    	Error:      	Received unexpected error:
                    	            	could not prefetch url: https://github.com/goreleaser/goreleaser/releases/download/v1.18.2/goreleaser_Darwin_arm64.tar.gz: exec: "nix-prefetch-url": executable file not found in $PATH: 
                    	Test:       	TestPrefetcher/prefetch/publish/valid
    --- FAIL: TestPrefetcher/available (0.00s)
        --- FAIL: TestPrefetcher/available/publish (0.00s)
            --- FAIL: TestPrefetcher/available/publish/valid (0.00s)
                nix_test.go:80: 
                    	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/nix/nix_test.go:80
                    	Error:      	Should be true
                    	Test:       	TestPrefetcher/available/publish/valid
...
--- FAIL: TestRun (0.91s)
    upx_test.go:119: 
        	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/upx/upx_test.go:119
        	Error:      	Received unexpected error:
        	            	upx not found in PATH
        	Test:       	TestRun
FAIL
FAIL	github.com/goreleaser/goreleaser/v2/internal/pipe/upx	1.728s
```

goreleaser's dev requirements are clearly documented in

https://github.com/goreleaser/goreleaser/blob/a9e0a8f11209b992b62e126274fb958e4527d65f/CONTRIBUTING.md?plain=1#L15-L23

However, requiring all of these raises the barrier to entry for contributors, and irrelevant failing tests add noise and time cost (they are re-run) for contributors who know which failures can be ignored.

### Describe the solution you'd like

Tests should be skipped if the required dependencies are not installed, for example:

```go
if _, err := os.LookPath("upx"); err != nil {
    t.Skip("upx not found in path")
}
```

Goreleaser's extensive CI will still catch problems, even if the PR author skips tests.

### Describe alternatives you've considered

Installing all the dependencies (painful: and some require non-trivial changes to the development machine, e.g. Docker).

Ignoring test failures (painful: the tests that fail due to my change are hard to spot amongst multiple other unrelated test failures).

Running `go test` just on the affected packages (painful: I need to work out which tests are relevant, and might get this wrong).

### Search

- [X] I did search for other open and closed issues before opening this

### Supporter

- [ ] I am a [sponsor](https://github.com/sponsors/caarlos0/) or a [Pro](https://goreleaser.com/pro) customer

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional context

I am happy to submit a PR for this. Just say!

--- a/internal/pipe/nix/nix_test.go
+++ b/internal/pipe/nix/nix_test.go
@@ -13,6 +13,7 @@ import (
 	"github.com/goreleaser/goreleaser/v2/internal/golden"
 	"github.com/goreleaser/goreleaser/v2/internal/skips"
 	"github.com/goreleaser/goreleaser/v2/internal/testctx"
+	"github.com/goreleaser/goreleaser/v2/internal/testlib"
 	"github.com/goreleaser/goreleaser/v2/pkg/config"
 	"github.com/stretchr/testify/require"
 )
@@ -35,6 +36,7 @@ func TestSkip(t *testing.T) {
 		}, testctx.Skip(skips.Nix))))
 	})
 	t.Run("nix-all-good", func(t *testing.T) {
+		testlib.CheckPath(t, "nix-prefetch-url")
 		require.False(t, NewPublish().Skip(testctx.NewWithCfg(config.Project{
 			Nix: []config.Nix{{}},
 		})))
@@ -62,6 +64,7 @@ func TestPrefetcher(t *testing.T) {
 				require.ErrorIs(t, err, exec.ErrNotFound)
 			})
 			t.Run("valid", func(t *testing.T) {
+				testlib.CheckPath(t, "nix-prefetch-url")
 				sha, err := publishShaPrefetcher{nixPrefetchURLBin}.Prefetch("https://github.com/goreleaser/goreleaser/releases/download/v1.18.2/goreleaser_Darwin_arm64.tar.gz")
 				require.NoError(t, err)
 				require.Equal(t, "0girjxp07srylyq36xk1ska8p68m2fhp05xgyv4wkcl61d6rzv3y", sha)
@@ -77,6 +80,7 @@ func TestPrefetcher(t *testing.T) {
 				require.False(t, publishShaPrefetcher{fakeNixPrefetchURLBin}.Available())
 			})
 			t.Run("valid", func(t *testing.T) {
+				testlib.CheckPath(t, "nix-prefetch-url")
 				require.True(t, publishShaPrefetcher{nixPrefetchURLBin}.Available())
 			})
 		})

--- a/internal/pipe/upx/upx_test.go
+++ b/internal/pipe/upx/upx_test.go
@@ -47,6 +47,7 @@ func TestSkip(t *testing.T) {
 }
 
 func TestRun(t *testing.T) {
+	testlib.CheckPath(t, "upx")
 	ctx := testctx.NewWithCfg(config.Project{
 		UPXs: []config.UPX{
 			{

